### PR TITLE
Add module X.Data.Conduit.Either

### PR DESCRIPTION
**Pull request by erikd-ambiata**
Conduit combinators for working with Conduits containing Either
values.

Code came from Jumbuck but need this functionality in Graft.

! @tmcgilchrist @jystic @nhibberd 
/jury approved @tmcgilchrist

### CODE REVIEW

File: x-conduit/ambiata-x-conduit.cabal [hunk 1]
```diff
@@ -44,12 +46,13 @@ test-suite test
...
-                     , text                            == 1.2.*
-                     , QuickCheck                      == 2.8.*
-                     , quickcheck-instances            == 0.3.*
+                     , bytestring                      == 0.10.*
                      , conduit                         == 1.2.*
                      , conduit-extra                   == 1.1.*
-                     , bytestring                      == 0.10.*
                      , filepath                        == 1.3.*
+                     , mtl                             == 2.2.*
+                     , QuickCheck                      == 2.8.*
```
Comment (tmcgilchrist) [inline x-conduit/ambiata-x-conduit.cabal]: This should probably be a range

File: x-conduit/src/X/Data/Conduit/Either.hs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+-- | Conduit component that allows the `Right` to be say written to a file
+-- while the `Left`s (eg warnings) are passed along the conduit.
+mapRightM :: MonadIO m => (a -> IO (Either e ())) -> Conduit (Either e a) m (Either e a)
+mapRightM actionR =
+  DC.awaitForever $
+    either (DC.yield . Left)
+        (\a -> liftIO (actionR a) >>= either (DC.yield . Left) (const $ DC.yield (Right a)))
+
+
+data ErrorCollectionParams = ErrorCollectionParams
```
Comment (tmcgilchrist) [inline x-conduit/src/X/Data/Conduit/Either.hs]: I wonder whether this belongs here or elsewhere (own library perhaps)? 
The functions above are nice and general, while the ones below are perhaps more specific to how we want to handle the eithers.

**Comment by tmcgilchrist** [comment]
This code exists in a few places now, makes sense to place it somewhere common.

**Comment by erikd-ambiata** [comment]
Dropped the error reporting stuff and added one general combinator.

**Comment by tmcgilchrist** [comment]
🚌

**Pull request merged**